### PR TITLE
BUG: DataFrame.diff(np_int_obj)

### LINKED ISSUE #44572: BUG: diff method raises ValueError when it's called with np.int64
### 

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the master branch of pandas.

### Reproducible Example

```python
import pandas as pd
import numpy as np

# w = np.array([2,3])[0]  # normal use 
w = np.int64(2)
pd.Series([1,2,3]).diff(w)     # works
pd.DataFrame([1,2,3]).diff(w)  # fails
```

### Issue Description

When `diff` method is called with a `np.int64` then `ValueError` is raised as  `np.int64` is not instance of `int` (see https://github.com/numpy/numpy/issues/2951#issuecomment-12747507). Also, this behavior is not consistent between `Series` and `DataFrame`.

### Expected Behavior

I've expected `DataFrame` work like `Series` does.

### Installed Versions

<details>

```
INSTALLED VERSIONS
------------------
commit           : 945c9ed766a61c7d2c0a7cbb251b6edebf9cb7d5
python           : 3.8.9.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.11.0-40-generic
Version          : #44~20.04.2-Ubuntu SMP Tue Oct 26 18:07:44 UTC 2021
machine          : x86_64
processor        :
byteorder        : little
LC_ALL           : None
LANG             : C.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.3.4
numpy            : 1.20.3
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 21.0.1
setuptools       : 56.0.0
Cython           : None
pytest           : 6.2.5
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : 7.29.0
pandas_datareader: None
bs4              : None
bottleneck       : 1.3.2
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.5.0
numexpr          : 2.7.3
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : 1.7.2
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : 0.54.1
```

</details>

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #44572
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

### CODE REVIEW

**Review by phofl — APPROVED** [review]
lgtm

File: pandas/core/frame.py [hunk 1]
```diff
@@ -8612,7 +8612,7 @@ def melt(
         ),
     )
     def diff(self, periods: int = 1, axis: Axis = 0) -> DataFrame:
-        if not isinstance(periods, int):
+        if not lib.is_integer(periods):
```
Comment (lithomas1) [inline pandas/core/frame.py:8615]: ```pandas/core/frame.py:8616: error: "int" has no attribute "is_integer"  [attr-defined]```

**Review by lithomas1 — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/frame.py:8615]: that's pretty weird, looks like a mypy problem to me

Comment (lithomas1) [reply pandas/core/frame.py:8615]: My bad, this actually refers to the line below ```if not (is_float(periods) and periods.is_integer()):```.

File: pandas/tests/frame/methods/test_diff.py [hunk 2]
```diff
@@ -17,6 +17,13 @@ def test_diff_requires_integer(self):
         with pytest.raises(ValueError, match="periods must be an integer"):
             df.diff(1.5)
 
+    def test_diff_allows_np_integer(self):
```
Comment (jreback) [inline pandas/tests/frame/methods/test_diff.py:20]: fine, could just parameterize the below one

**Review by jreback — APPROVED** [review]

File: pandas/tests/frame/methods/test_diff.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/tests/frame/methods/test_diff.py:20]: not wild about this as it is, but it looks like the next test could use a good splitting, after which combining might make more sense

**Review by lithomas1 — APPROVED** [review]

File: pandas/tests/frame/methods/test_diff.py (hunk 2 above)
Comment (jreback) [reply pandas/tests/frame/methods/test_diff.py:20]: kk add to list :->

**Pull request merged**